### PR TITLE
SLING-10489 : introducing partial startup detector

### PR DESCRIPTION
**Pull request by stefan-egli**

### CODE REVIEW

**Comment by stefan-egli** [comment]
PS1 : more test coverage to come
PS2 : depends on minor change in discovery.commons : https://github.com/apache/sling-org-apache-sling-discovery-commons/pull/3

File: src/test/java/org/apache/sling/discovery/oak/JoinerDelayTest.java [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        joinerDelay.cancelSync();
+    }
+
+    @Test
+    public void testSync_nulValues() throws Exception {
+        JoinerDelay joinerDelay = new JoinerDelay(1, scheduler);
+        try {
+            joinerDelay.sync(null, null);
+            fail("should fail");
+        } catch(Exception e) {
```
Comment (mreutegg) [inline src/test/java/org/apache/sling/discovery/oak/JoinerDelayTest.java:164]: I would catch and check for the more specific `IllegalArgumentException`.

File: src/test/java/org/apache/sling/discovery/oak/JoinerDelayTest.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            // ok
+        }
+        joinerDelay = new JoinerDelay(1, scheduler);
+        try {
+            joinerDelay.sync(view, null);
+            fail("should fail");
+        } catch(Exception e) {
```
Comment (mreutegg) [inline src/test/java/org/apache/sling/discovery/oak/JoinerDelayTest.java:171]: Same as above and there are a few more further down.

File: src/test/java/org/apache/sling/discovery/oak/JoinerDelayTest.java [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+            // ok
+        }
+        joinerDelay = new JoinerDelay(1, scheduler);
+        // this one is fine as it is caught
+        joinerDelay.sync(null, callback);
```
Comment (mreutegg) [inline src/test/java/org/apache/sling/discovery/oak/JoinerDelayTest.java:176]: I think the implementation should also throw an `IllegalArgumentException` when `view` is `null`. The contract does not mention `view` can be `null` and also does not specify what should happen in that case.

File: src/test/java/org/apache/sling/discovery/oak/OakDiscoveryServiceTest.java [hunk 4]
```diff
@@ -414,4 +428,352 @@ public void testPropertyProviderRegistrations() throws Exception{
...
+        VirtualInstance instance1 = builder1.build();
+        DummyListener listener1 = new DummyListener();
+        OakDiscoveryService discoveryService1 = (OakDiscoveryService) instance1.getDiscoveryService();
+        discoveryService1.bindTopologyEventListener(listener1);
+
+        instance1.heartbeatsAndCheckView();
+        instance1.heartbeatsAndCheckView();
+
+        assertEquals(0, discoveryService1.getViewStateManager().waitForAsyncEvents(2000));
+        Thread.sleep(1000);
```
Comment (mreutegg) [inline src/test/java/org/apache/sling/discovery/oak/OakDiscoveryServiceTest.java:460]: Why is this needed?

File: src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    private final boolean suppressingApplicable;
+    private final Set<Integer> partiallyStartedClusterNodeIds = new HashSet<>();
+
+    private final LogSilencer logSilencer;
+
+    /**
+     * @param lowestLocalSeqNum the lowest sequence number which
+     * the local OakClusterViewService has handled as part of asClusterView
+     * @param me the clusterNodeId (provided by oak) of the local instance (==me)
+     * @param localSlingIds slingId previously seen by this cluster instance (those will not be suppressed)
```
[the remark this answers is not in the document]
Comment (stefan-egli) [reply src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java:80]: That used to be the list of slingIds in the local cluster, already seen. It eventually moved to `OakClusterViewService.seenLocalInstances`.
I'll fix the javadoc here

File: src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        // * when the local instance ever showed to peers that it has fully started.
+        // and one way to verify for that is to demand that it ever wrote a synctoken.
+        // and to check that we keep note of the first ever successful seq num returned
+        // here
+        // and require the current syncToken to be at least that.
+        final long now = System.currentTimeMillis();
+        final long mySyncToken = readSyncToken(resourceResolver, mySlingId);
+        final boolean suppressionConfigured = config != null && config.getSuppressPartiallyStartedInstances();
+        suppressingApplicable = suppressionConfigured
+                && ((timeoutMillis <= 0) || (now < timeoutMillis))
```
[the remark this answers is not in the document]
Comment (stefan-egli) [reply src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java:107]: there should be 4 different cases:
1. suppression timeout is disabled (as is by default), in that case `timeoutMillis` is always 0 here, but suppressing should be applicable
2. there was never/no recent case of suppressing, `timeoutMillis` starts at 0 then, and suppressing should be applicable
3. with a suppression timeout configured and once the suppression starts, the `timeoutMillis` has a valid timestamp, in the future. Until then the suppressing is applicable
4. with a suppression timeout configured and a `timeoutMillis` of not 0, but a timestamp in the past the suppression no longer is applicable (the timeout feature).

I'll look into coding this a bit nicer

File: src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                && (mySyncToken >= lowestLocalSeqNum);
+        if (logger.isDebugEnabled()) {
+            logger.debug("<init> suppressionConfigured = " + suppressionConfigured + ", me = " + me + ", mySlingId = " + mySlingId +
+                    ", timeoutMillis = " + timeoutMillis + ", mySyncToken = " + mySyncToken +
+                    ", lowestLocalSeqNum = " + lowestLocalSeqNum + ", suppressingApplicable = " + suppressingApplicable);
+        }
+        this.logSilencer = logSilencer;
+    }
+
+    private boolean isSuppressing(int id) {
```
[the remark this answers is not in the document]
Comment (stefan-egli) [reply src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java:118]: agree, `isSuprressible()` is nicer, +1

File: src/test/java/org/apache/sling/discovery/oak/OakDiscoveryServiceTest.java (hunk 4 above)
Comment (stefan-egli) [reply src/test/java/org/apache/sling/discovery/oak/OakDiscoveryServiceTest.java:460]: `waitForAsyncEvents` could in theory have waited for an aysnc event in the queue to be picked up (by `AsyncEventSender.run()`), so `getInFlightEventCnt()` returns 0 - while that async event is still being executed via [asyncEvent.trigger()](https://github.com/apache/sling-org-apache-sling-discovery-commons/blob/8ecba4a098ae32e28022166a0e4095494efe10f1/src/main/java/org/apache/sling/discovery/commons/providers/base/AsyncEventSender.java#L118)

I think there's room for improvement here to avoid waiting - by more precise wait logic in the AsyncEventSender. Relying on sleeping is indeed not ideal.

File: src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (stefan-egli) [reply src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java]: done.

File: src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefan-egli) [reply src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java:112]: left the code as is but added docu

File: src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefan-egli) [reply src/main/java/org/apache/sling/discovery/oak/cluster/PartialStartupDetector.java]: done

File: src/test/java/org/apache/sling/discovery/oak/OakDiscoveryServiceTest.java (hunk 4 above)
Comment (stefan-egli) [reply src/test/java/org/apache/sling/discovery/oak/OakDiscoveryServiceTest.java:460]: Created SLING-10813 to follow-up. It would imply removing `Thread.sleep()` in many places - so ideally done outside of the scope of this PR, wdyt?

**Comment by stefan-egli** [comment]
@mreutegg , added a test case that reproduces your scenario : db70598681b76c4e6bb63fd2d35fa7b99ac6bcf0

**Comment by stefan-egli** [comment]
Updated the PR with the following change:
* disabled partial-startup-suppression by default to remain on the cautious side
* if enabled, the suppression now also applies to previous but slow members of the cluster : those cannot otherwise be distinguished from a restarting slow member (and either we suppress them or not). 
** the suggestion is instead to increase the minEventDelay to compensate for the removal of this strict check

**Comment by stefan-egli** [comment]
@mreutegg , now added a system property `org.apache.sling.discovery.oak.joinerdelay.enabled` that allows to overwrite the config as to whether this feature is enabled or not. The default of the config itself is disabled. The system property not being present hence results in the feature being disabled. This way IMO it is safe to go ahead with this PR, merge and release and revisit problems at a later stage if they arise. Wdyt?
